File: typescript/packages/mechanisms/evm/src/signer.ts
/**
 * ClientEvmSigner - Used by x402 clients to sign payment authorizations.
 *
 * Typically a viem WalletClient extended with publicActions:
 * ```typescript
 * const client = createWalletClient({
 *   account: privateKeyToAccount('0x...'),
 *   chain: baseSepolia,
 *   transport: http(),
 * }).extend(publicActions);
 * ```
 *
 * Or composed via `toClientEvmSigner(account, publicClient)`.
 */
export type ClientEvmSigner = {
  readonly address: `0x${string}`;
  signTypedData(message: {
    domain: Record<string, unknown>;
    types: Record<string, unknown>;
    primaryType: string;
    message: Record<string, unknown>;
  }): Promise<`0x${string}`>;
  /**
   * Optional on-chain reads.
   * Required only for extension enrichment (EIP-2612 / ERC-20 approval).
   */
  readContract?(args: {
    address: `0x${string}`;
    abi: readonly unknown[];
    functionName: string;
    args?: readonly unknown[];
  }): Promise<unknown>;
  /**
   * Optional: Signs a raw EIP-1559 transaction without broadcasting.
   * Required for ERC-20 approval gas sponsoring when the token lacks EIP-2612.
   */
  signTransaction?(args: {
    to: `0x${string}`;
    data: `0x${string}`;
    nonce: number;
    gas: bigint;
    maxFeePerGas: bigint;
    maxPriorityFeePerGas: bigint;
    chainId: number;
  }): Promise<`0x${string}`>;
  /**
   * Optional: Gets the current transaction count (nonce) for an address.
   * Required for ERC-20 approval gas sponsoring.
   */
  getTransactionCount?(args: { address: `0x${string}` }): Promise<number>;
  /**
   * Optional: Estimates current gas fees per gas.
   * Required for ERC-20 approval gas sponsoring.
   */
  estimateFeesPerGas?(): Promise<{ maxFeePerGas: bigint; maxPriorityFeePerGas: bigint }>;
};

/**
 * FacilitatorEvmSigner - Used by x402 facilitators to verify and settle payments
 * This is typically a viem PublicClient + WalletClient combination that can
 * read contract state, verify signatures, write transactions, and wait for receipts
 *
 * Supports multiple addresses for load balancing, key rotation, and high availability
 */
export type FacilitatorEvmSigner = {
  /**
   * Get all addresses this facilitator can use for signing
   * Enables dynamic address selection for load balancing and key rotation
   */
  getAddresses(): readonly `0x${string}`[];

  readContract(args: {
    address: `0x${string}`;
    abi: readonly unknown[];
    functionName: string;
    args?: readonly unknown[];
  }): Promise<unknown>;
  verifyTypedData(args: {
    address: `0x${string}`;
    domain: Record<string, unknown>;
    types: Record<string, unknown>;
    primaryType: string;
    message: Record<string, unknown>;
    signature: `0x${string}`;
  }): Promise<boolean>;
  writeContract(args: {
    address: `0x${string}`;
    abi: readonly unknown[];
    functionName: string;
    args: readonly unknown[];
    /** Optional gas limit. When provided, skips eth_estimateGas simulation. */
    gas?: bigint;
  }): Promise<`0x${string}`>;
  sendTransaction(args: { to: `0x${string}`; data: `0x${string}` }): Promise<`0x${string}`>;
  waitForTransactionReceipt(args: { hash: `0x${string}` }): Promise<{ status: string }>;
  getCode(args: { address: `0x${string}` }): Promise<`0x${string}` | undefined>;
};

/**
 * Composes a ClientEvmSigner from a local account and a public client.
 *
 * Use this when your signer (e.g., `privateKeyToAccount`) doesn't have
 * `readContract`. The `publicClient` provides the on-chain read capability.
 *
 * Alternatively, use a WalletClient extended with publicActions directly:
 * ```typescript
 * const signer = createWalletClient({
 *   account: privateKeyToAccount('0x...'),
 *   chain: baseSepolia,
 *   transport: http(),
 * }).extend(publicActions);
 * ```
 *
 * @param signer - A signer with `address` and `signTypedData` (and optionally `readContract`)
 * @param publicClient - A client with optional read/nonce/fee helpers
 * @param publicClient.readContract - The readContract method from the public client
 * @param publicClient.getTransactionCount - Optional getTransactionCount for ERC-20 approval
 * @param publicClient.estimateFeesPerGas - Optional estimateFeesPerGas for ERC-20 approval
 * @returns A ClientEvmSigner with any available optional capabilities
 *
 * @example
 * ```typescript
 * const account = privateKeyToAccount("0x...");
 * const publicClient = createPublicClient({ chain: baseSepolia, transport: http() });
 * const signer = toClientEvmSigner(account, publicClient);
 * ```
 */
export function toClientEvmSigner(
  signer: Omit<ClientEvmSigner, "readContract"> & {
    readContract?: ClientEvmSigner["readContract"];
  },
  publicClient?: {
    readContract(args: {
      address: `0x${string}`;
      abi: readonly unknown[];
      functionName: string;
      args?: readonly unknown[];
    }): Promise<unknown>;
    getTransactionCount?(args: { address: `0x${string}` }): Promise<number>;
    estimateFeesPerGas?(): Promise<{ maxFeePerGas: bigint; maxPriorityFeePerGas: bigint }>;
  },
): ClientEvmSigner {
  const readContract = signer.readContract ?? publicClient?.readContract.bind(publicClient);

  const result: ClientEvmSigner = {
    address: signer.address,
    signTypedData: msg => signer.signTypedData(msg),
  };

  if (readContract) {
    result.readContract = readContract;
  }

  // Forward optional capabilities from signer or publicClient
  const signTransaction = signer.signTransaction;
  if (signTransaction) {
    result.signTransaction = args => signTransaction(args);
  }

  const getTransactionCount =
    signer.getTransactionCount ?? publicClient?.getTransactionCount?.bind(publicClient);
  if (getTransactionCount) {
    result.getTransactionCount = args => getTransactionCount(args);
  }

  const estimateFeesPerGas =
    signer.estimateFeesPerGas ?? publicClient?.estimateFeesPerGas?.bind(publicClient);
  if (estimateFeesPerGas) {
    result.estimateFeesPerGas = () => estimateFeesPerGas();
  }

  return result;
}

/**
 * Converts a viem client with single address to a FacilitatorEvmSigner
 * Wraps the single address in a getAddresses() function for compatibility
 *
 * @param client - The client to convert (must have 'address' property)
 * @returns FacilitatorEvmSigner with getAddresses() support
 */
export function toFacilitatorEvmSigner(
  client: Omit<FacilitatorEvmSigner, "getAddresses"> & { address: `0x${string}` },
): FacilitatorEvmSigner {
  return {
    ...client,
    getAddresses: () => [client.address],
  };
}

File: python/x402/mechanisms/evm/signer.py
"""EVM signer protocol definitions."""

from typing import Any, Protocol, runtime_checkable

from .types import TransactionReceipt, TypedDataDomain, TypedDataField


class ClientEvmSigner(Protocol):
    """Client-side EVM signer for payment authorizations.

    Implement this protocol to integrate with your wallet provider
    (e.g., web3.py Account, eth-account, hardware wallet SDK).
    """

    @property
    def address(self) -> str:
        """The signer's Ethereum address (checksummed).

        Returns:
            Checksummed Ethereum address (0x...).
        """
        ...

    def sign_typed_data(
        self,
        domain: TypedDataDomain,
        types: dict[str, list[TypedDataField]],
        primary_type: str,
        message: dict[str, Any],
    ) -> bytes:
        """Sign EIP-712 typed data.

        Args:
            domain: EIP-712 domain separator.
            types: Type definitions.
            primary_type: Primary type name.
            message: Message data.

        Returns:
            65-byte ECDSA signature (r, s, v) or longer for smart wallets.
        """
        ...


class FacilitatorEvmSigner(Protocol):
    """Facilitator-side EVM signer for verification and settlement.

    Implement this protocol to integrate with your blockchain provider
    (e.g., web3.py, viem via adapter).
    """

    def get_addresses(self) -> list[str]:
        """Get all addresses this facilitator can use.

        Enables dynamic address selection for load balancing and key rotation.

        Returns:
            List of checksummed Ethereum addresses.
        """
        ...

    def read_contract(
        self,
        address: str,
        abi: list[dict[str, Any]],
        function_name: str,
        *args: Any,
    ) -> Any:
        """Read data from a smart contract.

        Args:
            address: Contract address.
            abi: Contract ABI.
            function_name: Function to call.
            *args: Function arguments.

        Returns:
            Function return value.
        """
        ...

    def verify_typed_data(
        self,
        address: str,
        domain: TypedDataDomain,
        types: dict[str, list[TypedDataField]],
        primary_type: str,
        message: dict[str, Any],
        signature: bytes,
    ) -> bool:
        """Verify an EIP-712 signature.

        Args:
            address: Expected signer address.
            domain: EIP-712 domain separator.
            types: Type definitions.
            primary_type: Primary type name.
            message: Message data.
            signature: Signature bytes.

        Returns:
            True if signature is valid.
        """
        ...

    def write_contract(
        self,
        address: str,
        abi: list[dict[str, Any]],
        function_name: str,
        *args: Any,
    ) -> str:
        """Execute a smart contract transaction.

        Args:
            address: Contract address.
            abi: Contract ABI.
            function_name: Function to call.
            *args: Function arguments.

        Returns:
            Transaction hash.
        """
        ...

    def send_transaction(self, to: str, data: bytes) -> str:
        """Send raw transaction.

        Args:
            to: Recipient address.
            data: Transaction data.

        Returns:
            Transaction hash.
        """
        ...

    def wait_for_transaction_receipt(self, tx_hash: str) -> TransactionReceipt:
        """Wait for transaction to be mined.

        Args:
            tx_hash: Transaction hash to wait for.

        Returns:
            Transaction receipt.
        """
        ...

    def get_balance(self, address: str, token_address: str) -> int:
        """Get token balance for address.

        Args:
            address: Account address.
            token_address: Token contract address.

        Returns:
            Balance in smallest unit.
        """
        ...

    def get_chain_id(self) -> int:
        """Get connected network's chain ID.

        Returns:
            Chain ID.
        """
        ...

    def get_code(self, address: str) -> bytes:
        """Get bytecode at address.

        Args:
            address: Address to check.

        Returns:
            Bytecode (empty if EOA).
        """
        ...


@runtime_checkable
class ClientEvmSignerWithReadContract(Protocol):
    """Extension of ClientEvmSigner that adds on-chain read capability.

    Required for EIP-2612 gas sponsoring (needs to read nonces from token).
    """

    @property
    def address(self) -> str: ...

    def sign_typed_data(
        self,
        domain: TypedDataDomain,
        types: dict[str, list[TypedDataField]],
        primary_type: str,
        message: dict[str, Any],
    ) -> bytes: ...

    def read_contract(
        self,
        address: str,
        abi: list[dict[str, Any]],
        function_name: str,
        *args: Any,
    ) -> Any:
        """Read data from a smart contract."""
        ...


@runtime_checkable
class ClientEvmSignerWithSignTransaction(Protocol):
    """Extension of ClientEvmSigner that adds raw transaction signing.

    Required for ERC-20 approval gas sponsoring (signs approve tx off-chain).
    """

    @property
    def address(self) -> str: ...

    def sign_typed_data(
        self,
        domain: TypedDataDomain,
        types: dict[str, list[TypedDataField]],
        primary_type: str,
        message: dict[str, Any],
    ) -> bytes: ...

    def sign_transaction(self, tx: dict[str, Any]) -> str:
        """Sign an EIP-1559 transaction and return the RLP-encoded hex string."""
        ...

    def get_transaction_count(self, address: str) -> int:
        """Get the pending nonce for an address."""
        ...
